test/e2e/scale.go

```diff
@@ -62,96 +63,108 @@ func ScaleToWithin(t *testing.T, scale int, duration time.Duration, latencies La
 				}
-				url = s.Status.URL.URL()
-				return v1a1test.IsServiceReady(s)
-			}, "ServiceUpdatedWithURL")
-			if err != nil {
-				t.Errorf("WaitForServiceState(w/ Domain) = %v", err)
-				return fmt.Errorf("WaitForServiceState(w/ Domain) failed: %w", err)
-			}
-			// Record the time it took to become ready.
-			latencies.Add("time-to-ready", start)
-
-			_, err = pkgTest.WaitForEndpointState(
-				clients.KubeClient,
-				t.Logf,
-				url,
-				v1a1test.RetryingRouteInconsistency(pkgTest.MatchesAllOf(pkgTest.IsStatusOK, pkgTest.MatchesBody(test.HelloWorldText))),
-				"WaitForEndpointToServeText",
-				test.ServingFlags.ResolvableDomain)
-			if err != nil {
-				t.Errorf("WaitForEndpointState(expected text) = %v", err)
-				return fmt.Errorf("WaitForEndpointState(expected text) failed: %w", err)
-			}
-			// Record the time it took to get back a 200 with the expected text.
-			latencies.Add("time-to-200", start)
+				// Record the time it took to create the service.
+				latencies.Add("time-to-create", start)
+				names.Route = serviceresourcenames.Route(svc)
+				names.Config = serviceresourcenames.Configuration(svc)
+
+				// Send it to our cleanup logic (below)
+				cleanupCh <- names
+
+				t.Logf("Wait for %s to become ready.", names.Service)
+				var url *url.URL
+				err = v1a1test.WaitForServiceState(clients.ServingAlphaClient, names.Service, func(s *v1alpha1.Service) (bool, error) {
+					if s.Status.URL == nil {
+						return false, nil
+					}
+					url = s.Status.URL.URL()
+					return v1a1test.IsServiceReady(s)
+				}, "ServiceUpdatedWithURL")
+				if err != nil {
+					t.Errorf("WaitForServiceState(w/ Domain) = %v", err)
+					return fmt.Errorf("WaitForServiceState(w/ Domain) failed: %w", err)
+				}
+				// Record the time it took to become ready.
+				latencies.Add("time-to-ready", start)
+
+				_, err = pkgTest.WaitForEndpointState(
+					clients.KubeClient,
+					t.Logf,
+					url,
+					v1a1test.RetryingRouteInconsistency(pkgTest.MatchesAllOf(pkgTest.IsStatusOK, pkgTest.MatchesBody(test.HelloWorldText))),
+					"WaitForEndpointToServeText",
+					test.ServingFlags.ResolvableDomain)
+				if err != nil {
+					t.Errorf("WaitForEndpointState(expected text) = %v", err)
+					return fmt.Errorf("WaitForEndpointState(expected text) failed: %w", err)
+				}
+				// Record the time it took to get back a 200 with the expected text.
+				latencies.Add("time-to-200", start)
 
-			// Start probing the domain until the test is complete.
-			pm.Spawn(url)
+				// Start probing the domain until the test is complete.
+				pm.Spawn(url)
 
-			t.Logf("%s is ready.", names.Service)
-			return nil
+				t.Logf("%s is ready.", names.Service)
+				return nil
+			}
+			errCh := make(chan error)
+			ctx, cancel := context.WithTimeout(context.Background(), duration)
+			defer cancel()
+			go func() {
+				errCh <- worker()
+			}()
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("Timed out waiting for service")
```

Comment:
This would still cause the goroutine to dangle around and potentially cause the panics that you mentioned. Is this really different to what we had here before?